pkg/trace/sampler/exception_sampler.go

```diff
@@ -0,0 +1,210 @@
+package sampler
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// cardinalityLimit limits the number of spans considered per combination of (env, service).
+	cardinalityLimit = 1000
+	// defaultTTL limits the frequency at which we sample a same span (env, service, name, rsc, ...).
+	defaultTTL = 2 * time.Minute
+	// priorityTTL allows to blacklist p1 spans that are sampled entirely.
+	priorityTTL = 10 * time.Minute
+	// ttlRenewalPeriod prevents from continuously updating expire times for each span seen.
+	ttlRenewalPeriod = 1 * time.Minute
+	// exceptionSamplerTPS traces per second allowed by the rate limiter.
+	exceptionSamplerTPS = 5
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
+// It ensures that we sample traces for each combination of
+// (env, service, name, resource, error type, http status) seen on a top level or measured span
+// for which we did not see any span with a priority > 0 (sampled by Priority).
+// The resulting sampled traces will likely be incomplete.
+type ExceptionSampler struct {
+	mu      sync.RWMutex
+	limiter *rate.Limiter
+	seen    map[Signature]*seenSpans
+
+	tickStats *time.Ticker
+	hits      int64
+	misses    int64
+	shrinks   int64
+}
+
+// NewExceptionSampler returns a NewExceptionSampler that ensures that we sample combinations
+// of env, service, name, resource, http-status, error type for each top level or measured spans
+func NewExceptionSampler() *ExceptionSampler {
+	return &ExceptionSampler{
+		limiter:   rate.NewLimiter(exceptionSamplerTPS, exceptionSamplerBurst),
+		seen:      make(map[Signature]*seenSpans),
+		tickStats: time.NewTicker(10 * time.Second),
+	}
+}
+
+// Add samples a trace and returns true if trace was sampled (should be kept)
+func (e *ExceptionSampler) Add(env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	return e.add(time.Now(), env, root, t)
+}
+
+func (e *ExceptionSampler) add(now time.Time, env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	if priority, ok := GetSamplingPriority(root); priority > 0 && ok {
+		e.handlePriorityTrace(now, env, t)
+		return false
+	}
+	return e.handleTrace(now, env, t)
+}
+
+// Start starts reporting stats
+func (e *ExceptionSampler) Start() {
+	go func() {
+		for range e.tickStats.C {
+			e.report()
+		}
+	}()
```

Comment:
Can we move this code into `NewExceptionSampler` and have only `Stop` ? It feels weird to have to call `New` all the time followed by `Start`. If that complicates your test code or prolongs things unnecessarily, it's fine to leave it for now.